lib/sidekiq/job/iterable.rb

```diff
@@ -0,0 +1,245 @@
+          end
+
+          super
+        end
+      end
+
+      attr_reader :executions,
+        :cursor_position,
+        :times_interrupted
+      # The time when the job starts running. If the job is interrupted and runs again,
+      # the value is updated.
+      attr_reader :start_time
+
+      # The total time the job has been running, including multiple iterations.
+      # The time isn't reset if the job is interrupted.
+      attr_reader :total_time
+
+      # @private
+      def initialize
+        super
+        @arguments = nil
+        @needs_reenqueue = false
+      end
+
+      # @private
+      def perform(*arguments)
+        extract_previous_runs_metadata(arguments)
+        @arguments = arguments
+        interruptible_perform(*arguments)
+      end
+
+      # A hook to override that will be called when the job starts iterating.
+      # Is called only once, for the first time.
+      def on_start
+      end
+
+      # A hook to override that will be called around each iteration.
+      # Can be useful for some metrics collection, performance tracking etc.
+      def around_iteration
+        yield
+      end
+
+      # A hook to override that will be called when the job resumes iterating.
+      def on_resume
+      end
+
+      # A hook to override that will be called each time the job is interrupted.
+      # This can be due to throttling (throttle enumerator), `max_job_runtime` configuration,
+      # or sidekiq restarting.
+      def on_shutdown
+      end
+
+      # A hook to override that will be called when the job finished iterating.
+      def on_complete
+      end
+
+      # The enumerator to be iterated over.
+      #
+      # @return [Enumerator]
+      #
+      # @raise [NotImplementedError] with a message advising subclasses to
+      #     implement an override for this method.
+      #
+      def build_enumerator(*)
+        raise NotImplementedError, "#{self.class.name} must implement a 'build_enumerator' method"
+      end
+
+      # The action to be performed on each item from the enumerator.
+      #
+      # @return [void]
+      #
+      # @raise [NotImplementedError] with a message advising subclasses to
+      #     implement an override for this method.
+      #
+      def each_iteration(*)
+        raise NotImplementedError, "#{self.class.name} must implement an 'each_iteration' method"
+      end
+
+      def throttle?
+        false
+      end
+
+      private
+        def extract_previous_runs_metadata(arguments)
+          options =
+            if arguments.last.is_a?(Hash) && arguments.last.key?("sidekiq_iteration")
+              arguments.pop["sidekiq_iteration"]
+            else
+              {}
+            end
+
+          @executions = options["executions"] || 0
+          @cursor_position = options["cursor_position"]
+          @times_interrupted = options["times_interrupted"] || 0
+          @total_time = options["total_time"] || 0
+        end
```

Comment:
The iterable job keeps its state inside the job hash metadata itself (as the last item of the args, if to be concrete). So, if the interrupted job is run again, for example, we restore its attributes from the previously stored metadata. Same when the job raises and error - we save the current state into the job and push it to redis.

Another approach is to keep the iteration state as a separate structure in redis (as a hash, e.g.). But then we need to decide on expiration time of that hash, decide if we should update it after each iteration (+1 redis call per iteration) or once in a while, write a server middleware to set job's attributes based on that hash.

Which one would you suggest/prefer?

Reply:
That's a really tough question. It's pretty fundamental to the design and I'm not sure I like the idea of mutating the job payload itself, as it's stored in Redis as a String which can't easily be modified without a JSON parse/dump round trip. I'm leaning toward defining a job iteration Hash, with a key like "it-{jid}". Redis has a number of useful H* commands to update/increment Hash values very quickly.

I think the default should be to update the iteration record in Redis only if the iteration is interrupted or 5 seconds have passed. It should not be updated after every record; if we are processing a million records we don't want a million Redis updates.